python/ray/data/datasource/tfrecords_datasource.py

```diff
@@ -0,0 +1,116 @@
+                    f"message in '{path}'. This warning can occur if your TFRecord "
+                    "file contains a message type other than `tf.train.Example`."
+                )
+
+            data.append(_convert_example_to_dict(example))
+
+        return pd.DataFrame.from_records(data)
+
+
+def _convert_example_to_dict(
+    example: tf.train.Example,
+) -> Dict[str, Union[bytes, List[bytes], float, List[float], int, List[int]]]:
+    record = {}
+    for feature_name, feature in example.features.feature.items():
+        value = _get_feature_value(feature)
+        if len(value) == 1:
+            value = value[0]
+        record[feature_name] = value
+    return record
+
+
+def _get_feature_value(
+    feature: tf.train.Feature,
+) -> Union[List[bytes], List[float], List[int]]:
+    values = (
+        feature.bytes_list.value,
+        feature.float_list.value,
+        feature.int64_list.value,
+    )
+    # Exactly one of `bytes_list`, `float_list`, and `int64_list` should contain data.
+    assert sum(bool(value) for value in values) == 1
+
+    if feature.bytes_list.value:
+        return list(feature.bytes_list.value)
+    if feature.float_list.value:
+        return list(feature.float_list.value)
+    if feature.int64_list.value:
+        return list(feature.int64_list.value)
+
+
+# Adapted from https://github.com/vahidk/tfrecord/blob/master/tfrecord/reader.py#L16-L96
+#
+# MIT License
+#
+# Copyright (c) 2020 Vahid Kazemi
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in all
+# copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+def _read_records(
+    file: "pyarrow.NativeFile",
+) -> Iterable[memoryview]:
+    length_bytes = bytearray(8)
+    crc_bytes = bytearray(4)
+    datum_bytes = bytearray(1024 * 1024)
+    while True:
+        num_length_bytes_read = file.readinto(length_bytes)
+        if num_length_bytes_read == 0:
+            break
+        elif num_length_bytes_read != 8:
+            raise RuntimeError("Failed to read the record size.")
```

Comment:
May want to use a more specific exception type for these, I think that `ValueError` would be more appropriate than `RuntimeError`.